src/fitutil/SystematicManager.cpp

```diff
@@ -1,39 +1,196 @@
 #include <SystematicManager.h>
+#include <Exceptions.h>
+#include <Formatter.hpp>
+#include <ContainerTools.hpp>
 
-const std::vector<Systematic*>& 
-SystematicManager::GetSystematics() const{
-    return fSystematics;
-}
 
 void
 SystematicManager::Construct(){
-    // Dont do anything if there are no systematics
-    if(!fSystematics.size())
+    // Don't do anything if there are no systematics
+    if(!fGroups.size())
         return;
 
-    // construct the response matricies
-    for(size_t i = 0; i < fSystematics.size(); i++)
-        fSystematics[i] -> Construct();
+    //Construct the response matrices.
+    for (std::map<std::string,std::vector<Systematic*> >::const_iterator group=fGroups.begin(); group!= fGroups.end(); ++group ) {
+        for(size_t i = 0; i < group->second.size(); i++)
+            fGroups[group->first].at(i) -> Construct();
+    }
 
-    
-    // Assemble the overall response
-    fTotalResponse = fSystematics.at(0) -> GetResponse();
-    for(size_t i = 1; i < fSystematics.size(); i++)
-      fTotalResponse *= fSystematics.at(i) -> GetResponse();
+    //This loop should construct all fGroups.
+    for (std::map<std::string,std::vector<Systematic*> >::const_iterator group =fGroups.begin(); group != fGroups.end(); ++group ) {
+        SparseMatrix resp = fGroups[group->first].at(0) -> GetResponse();
+        for(size_t i = 1; i < group->second.size(); i++)
+            resp *= fGroups[group->first].at(i) -> GetResponse();
+        fTotalReponses[group->first]=resp;
+    }
 }
 
 const SparseMatrix&
-SystematicManager::GetTotalResponse() const{
-     return fTotalResponse;
+SystematicManager::GetTotalResponse(const std::string& groupName_) const{
+    //should you construct?
+    try{
+        return fTotalReponses.at(groupName_);
+    }
+    catch(const std::out_of_range& e_){
+        throw NotFoundError(Formatter()<<
+                "SystematicManager :: name : "<< groupName_<<
+                " not found in systematic map");
+    }
+}
+
+
+void
+SystematicManager::Add(Systematic* sys_, const std::string& groupName_){
+    fGroups[groupName_].push_back(sys_);
+    fNGroups = fGroups.size();
+}
+
+void
+SystematicManager::UniqueSystematics(const std::vector<std::string>& syss_){
+    std::vector<std::string> allname;
+    for (int i = 0; i <syss_.size(); ++i) {
+        if(fGroups.find( syss_.at(i) ) == fGroups.end())
+            throw NotFoundError (
+                    Formatter()<<"SystematicManager:: Systematic group "<<
+                    syss_.at(i)<<
+                    " not found in complete set of groups. " );
```

Comment:
Might be worth rewording this one? I spent a while trying to work out what a complete set of groups was before I realised I was being dim.